### ISSUE TITLE
External import maps support

### ISSUE THREAD

**Issue by guybedford** [WICG/import-maps#235]
Currently external import maps are specified here, but in Chrome these still give the error:

```
External import maps are not yet supported.
```

I think it could be useful to clearly indicate in the spec that this is an optional feature of the implementation here, or possibly even split it out, given the current movements - especially when other browsers are being asked to consider implementation work.

**Comment by domenic** [comment]
We're hoping to implement ASAP, although there may be a 6-week period where they aren't implemented.

**Comment by guybedford** [comment]
Ok, if that's the plan then this can be closed certainly.

**Comment by LarsDenBakker** [comment]
What's the current planning on this?

**Comment by domenic** [comment]
@hiroshige-g, how's the external import maps implementation going?

**Comment by guybedford** [comment]
@domenic thinking about the performance tradeoff here - does encouraging speculative fetching avoid the problem?

**Comment by joeldenning** [comment]
I put together some thoughts and a benchmark in this document about this topic: https://docs.google.com/document/d/1IxkNGRIUbvQ0P8Uig9jkFte5gACCn-Y9zO1P5acnE9s/edit#

**Comment by joeldenning** [comment]
One of my main points in that document:

> External import maps can be controlled and cached separately from the HTML files that reference them, which allows for updating javascript resources without busting the cache for the main HTML resource.

**Comment by canonic-epicure** [comment]
I believe "external" import maps of course needs to be supported. Its the developer's decision, which one to choose.

Just stumbled upon this issue. I thought Chrome supports import maps fully. Just when, oh when, web will finally switch to bunder-free development?

Spend whole day yesterday trying to make Snowpack to work with my app, trying Vite today. Keep filing github issues. All I need is a simple bare specifier support in few places.

**Comment by davidmarkclements** [comment]
hey everyone just bumping on this - do we have an updated eta on external import maps anywhere?

**Comment by bartlomieju** [comment]
FWIW Deno has an `--import-map` flag, that currently accepts a file path or a URL that allows to load both local and remote import map. I know this is not directly linked to the issue at hand but we found ability to load remote import maps useful. Thumbs up from me.

**Comment by Brain2000** [comment]
I'm having an issue trying to understasnd why it would be that much work to make it work externally, like every other script tag works internal or external. The only difference here is verifying the content-type of application/importmap+json being returned. It's not reinventing the wheel.... unless chrome is a pile of spaghetti and it is reinventing the wheel.

**Comment by Brain2000** [comment]
This isn't "formally" an external importmap, but it works:

test.htm:
```
<html>
<head>
<script src="./test.js"></script>

<script type="module">
import 'my-library'; // will fetch the randomly-chosen URL
</script>

</head>
<body>
</body>
</html>
```

test.js:
```
const im = document.createElement('script');
im.type = 'importmap';
im.textContent = JSON.stringify({
  imports: {
    'my-library': Math.random() > 0.5 ? '/my-awesome-library.mjs' : '/my-rad-library.mjs'
  }
});
document.currentScript.after(im);
```

**Comment by Brain2000** [comment]
I've also found that this does not work with importmaps, though I should probably open a seaprate issue for this:

```
<link rel="modulepreload" href="my-library">
```
This always loads "my-library" verbatim instead of what the importmap depicts.
This breaks some bundlers, like vite, which injects modulepreload for every chunk, and it has to be stripped.

**Comment by masx200** [comment]
```json

{
  "imports": {
    "a": "/a-1.mjs",
    "b": "/b-1.mjs",
    "c": "/c-1.mjs"
  },
  "scopes": {
    "/scope2/": {
      "a": "/a-2.mjs"
    },
    "/foo/": "/foo.map.json"
  }
}
```

**Comment by guybedford** [comment]
@masx200 specifically defining scope boundaries for externalization is a nice thought! That said, import map resolutions cannot be individually asynchronous therefore allowing specific resolutions to trigger and wait on async import map fetching work is not possible.

**Comment by masx200** [comment]
```json
{
"externals":{
"http://example.com/foo/":"http://example.com/foo/import_map.json"

},
  "imports": {
    "a": "http://example.com/a/mod.mjs",
    "scope2/": "http://example.com/scope2/",
    "foo/":"http://example.com/foo/"
  },
  "scopes": {
    "http://example.com/scope2/": {
      "b": "http://example.com/b/mod.mjs"
    }
    
  }
}
```

**Comment by FlippieCoetser** [comment]
Has any decision been made on whether external import maps will be supported? 

I am using Importmaps as in the below example:
```html
<!doctype html>
<html>
  <head>
    <title>Demo</title>
    <link rel="icon" href="favicon.ico" type="image/x-icon">
    <link rel="stylesheet" href="html.style.css">
    <script type="importmap">
      {
          "imports": {
            "@browser-modules/dictionary": "../node_modules/@browser-modules/dictionary/lib/dictionary.js",
            "@browser-modules/events": "../node_modules/@browser-modules/events/lib/events.js",
            "@browser-modules/component.library": "../node_modules/@browser-modules/component.library/lib/library.js",	
            "@browser-modules/machine": "../node_modules/@browser-modules/machine/lib/machine.js",
            "@browser-modules/web.component": "../node_modules/@browser-modules/web.component/lib/component.js",
            "@browser-modules/buttons": "../node_modules/@browser-modules/buttons/lib/buttons.js"
          }
      }
    </script>
  </head>
  <body>
    <script src="index.js" type="module"></script>
    <task-bar></task-bar>
  </body>
</html>
``` 

However, I realize that the list of imports will soon be extensive, and instead, I prefer to use external import maps. 

**Can anyone tell me if this GitHub issue will be addressed?**

**Comment by pookage** [comment]
> We're hoping to implement ASAP, although there may be a 6-week period where they aren't implemented.

I am LOVING the optimism of this 3-year old comment 😂

Just adding my voice those eagerly awaiting this feature 👀

**Comment by FluorescentHallucinogen** [comment]
As an idea, what about a simple polyfill that takes external import map file and inline its content into HTML code? This would allow developers to use external import maps now. This could be implemented both in runtime and at build time (e.g. as a Rollup plugin).

**Comment by FlippieCoetser** [comment]
@FluorescentHallucinogen I believe the challenge with the approach you outline is that any javascript that fetches an external file runs async, which runs only after the importmap has been loaded. Ideally, we need support by default for external importmap. Or am I missing something here?

**Comment by FlippieCoetser** [comment]
@domenic, I noticed that in the second comment in the list, you provided an estimated timeline. Since it is a rather old comment, could you share your thoughts on when external import maps will be supported? No pressure; I want to calibrate my expectations.  Such functionality will really be helpful.

**Comment by domenic** [comment]
I have no estimate at this time.

**Comment by FluorescentHallucinogen** [comment]
@FlippieCoetser
> I believe the challenge with the approach you outline is that any javascript that fetches an external file runs async, which runs only after the importmap has been loaded.

Not sure about run-time polyfill, but it's definitely possible at build-time (including serving project while developing). I've developed a dirty proof-of-concept, and it works! See https://github.com/modernweb-dev/web/issues/2562#issuecomment-1817907520 for more info. 😉

All tools I know (https://github.com/jsenv/importmap-node-module, https://www.npmjs.com/package/importly) generate import maps as external files.

BTW, maybe `nginx` can be configured somehow to inline/inject import map from an external file into the `index.html` code on the fly. 🤔

**Comment by nenadvicentic** [comment]
@FluorescentHallucinogen, what you are describing, is not browser polyfil. Your build-time polyfill is a build step, tied to a specific development environment and perhaps specific type of projects and/or project structure. Browser polyfill should be independent of backends and backend tools.

It is good to know that such tools exist, but let's not mix up build transformation step and discussion about general browser support of a feature.

**Comment by hjabbing** [comment]
As a workaround, I point the import value to an index.js file where I re-export the modules (and versions) I want. That way I don't need to modify the page with the importmap which is out of my control, only the index.js which I do have control over. 
This works but please let me know if I'm missing something.

**Comment by bitifet** [comment]
Regarding performance issues I understand that external import maps require an extra network request which may imply an unnecessary delay in SPAs

But in the case of multi page applications, from my point of view, that small delay in the landing page will be so worth having the rest of pages the user visit will just get it from cache...

I'm aware that nowadays most developers are tied to one or another framework all of them SPA based*, but MPA's do exist too and, in my opinion, they deserve our attention too.

(*) I myself love the SPA concept even I believe a mix of both worlds is possible (through multiple entry point SPAs and History API), but that's another history...


Anyway, don't take me wrong: I'm not voting for or against native support of external imports: I understand it could be a noteworthy amount of work and, hence, it's important to know if it's going to be useful to so many people beforehand.

But, even it can be polyfilled, it would imply the use of a bundler or, at least, some manual setup.

And bundlers are great, but they require an extra setup and maintenance effort and not all web applications are Facebook or Twitter. There are also lots of small apps that deserve to see the light and many times they doesn't just due to the time we need to invest in their initial setup.

In fact I got here while investigating the convenience of adding support for external import maps to [esmrouter](https://www.npmjs.com/package/esmrouter) which is a small package I implemented to ease making all [browser-enabled npm packages](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#browser) directly available client side in Express applications.

And, by the way, it works perfectly with inlined import maps. But I still feel like there's something missing... 😉

**Comment by canonic-epicure** [comment]
The external import maps are the final step towards the bundlers-free web development. Such a shame, that step is still not done. After ~20 years of web/js existence, you still can not install your dependencies and run your code directly, w/o passing it through the bundlers "meat grinder".

The workflow that will hopefully exists one day, is simple: you run `npm i` and in your setup, you specify to run some script after every `npm i` that will scan all deps and generate an import map file and store it near the `package.json`. Then in your web page, you just include that import map - and you can use your dependencies immediately. 

Bundlers should only be used for creating a production build of the app. Using bundlers for development is a legacy practice from the CommonJS times, which still dominates because of the lacking imports map support.

**Comment by klebba** [comment]
Currently it seems that import maps simply will not work within a Chrome Extension context until this is addressed.

**Comment by guybedford** [comment]
@klebba can you explain further the chrome extension issue here?

**Comment by klebba** [comment]
Surprisingly it does not seem to be possible to declare an `importmap` in a Manifest v3 browser extension context. Inlining the `importmap` is disallowed by intrinsic CSP rules (even though the script content is JSON). Modifying the `innerHTML` or `textContent` of a `<script>` tag is also disallowed by the intrinsic CSP rules in Manifest v3. In seeking alternatives I found this perpetually disappointing thread.

*There is apparently no method for using an import map that functions today in a Manifest v3 Chrome Extension.*

Related SO thread: https://stackoverflow.com/a/78557452


Oct 2020:
> We're hoping to implement ASAP, although there may be a 6-week period where they aren't implemented.
[link](https://github.com/WICG/import-maps/issues/235#issuecomment-719766779)

Oct 2021:
> Nobody on the Chrome team is currently assigned to work on import maps.
[link](https://github.com/WICG/import-maps/issues/235#issuecomment-943728519)

May 2024:
> Manifest V2 phase-out begins
[link](https://blog.chromium.org/2024/05/manifest-v2-phase-out-begins.html)

I guess we should expect the Chrome team will imminently implement this feature or risk enduring _ever more_ thumbs up here?

**Comment by devingfx** [comment]
+1 for the pain of chrome extension CSP !

**Comment by joeldenning** [comment]
For anybody looking for a workaround: [import-map-injector](https://github.com/single-spa/import-map-injector?tab=readme-ov-file#import-map-injector) is a 3kb npm package that supports external import maps by downloading them and then injecting them as a native inline import map

**Comment by stopsopa** [comment]
It seems [Multiple import maps](https://chromestatus.com/feature/5121916248260608) was just added to chromium version 133
